File: 1282-113-1408-string-matching-in-an-array/1282-113-1408-string-matching-in-an-array.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.freq = 0

class Solution:
    def __init__(self):
        self.trie = TrieNode()

    def insert(self, word):
        node = self.trie
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
            node.freq += 1

    def get(self, word):
        node = self.trie
        for ch in word:
            node = node.children[ch]
        return node.freq > 1

    def stringMatching(self, words: list[str]) -> list[str]:
        for word in words:
            for j in range(len(word)):
                self.insert(word[j:])
        
        result = []
        for word in words:
            if self.get(word):
                result.append(word)
        
        return result

File: 1282-113-1408-string-matching-in-an-array/1282-113-1408-string-matching-in-an-array.cpp
class Solution {
public:
    struct Trie{
        Trie* vis[26];
        int frq = 0;
    }*trie;

    void insert(string s){
        Trie * top = trie;
        for(char ch : s){
            if(!top->vis[ch-'a'])
                top->vis[ch-'a'] = new Trie();
            top = top->vis[ch-'a'];
            top->frq++;
        }
    }
    bool get(string s){
        Trie* top = trie;
        for(char ch : s)
            top = top->vis[ch-'a'];
        
        return top->frq > 1;
    }

    vector<string> stringMatching(vector<string>& words) {
        trie = new Trie();
        for(int i = 0 ; i < words.size();i++)
            for(int j = 0 ; j < words[i].size();j++)
                insert(words[i].substr(j));
        
            
        vector<string>v;
        for(int i = 0 ; i < words.size();i++)
            if(get(words[i]))v.push_back(words[i]);
        
        return v;
    }
};